File: examples/android/BibleApiClient.kt
// ============================================================
// Bible Mirror API — Android Kotlin Client
// ============================================================
// build.gradle.kts dependencies:
//   implementation(platform("com.google.firebase:firebase-bom:32.7.0"))
//   implementation("com.google.firebase:firebase-auth-ktx")
//   implementation("com.squareup.okhttp3:okhttp:4.12.0")
// ============================================================

package com.yourapp.bible.api

import com.google.firebase.auth.FirebaseAuth
import kotlinx.coroutines.tasks.await
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.io.IOException

object BibleApiConfig {
    // Ganti dengan URL Worker Anda setelah deploy
    const val BASE_URL = "https://bible-mirror-api.<your-subdomain>.workers.dev"
}

data class ApiMeta(val cached: Boolean, val timestamp: String, val version: String)

sealed class ApiResult<out T> {
    data class Success<T>(val data: T, val meta: ApiMeta) : ApiResult<T>()
    data class Error(val code: String, val message: String, val status: Int) : ApiResult<Nothing>()
    data class NetworkError(val throwable: Throwable) : ApiResult<Nothing>()
}

class BibleApiClient(
    private val auth: FirebaseAuth = FirebaseAuth.getInstance(),
    private val http: OkHttpClient = OkHttpClient(),
) {
    private suspend fun getIdToken(): String {
        val user = auth.currentUser ?: throw IllegalStateException("User not signed in")
        return user.getIdToken(false).await().token ?: throw IllegalStateException("Token null")
    }

    private suspend fun get(path: String): ApiResult<JSONObject> {
        val token = try { getIdToken() } catch (e: Exception) {
            return ApiResult.Error("UNAUTHORIZED", e.message ?: "Auth failed", 401)
        }
        val request = Request.Builder()
            .url("${BibleApiConfig.BASE_URL}$path")
            .header("Authorization", "Bearer $token")
            .header("Accept", "application/json")
            .build()
        return try {
            val response = http.newCall(request).execute()
            val body = response.body?.string() ?: "{}"
            val json = JSONObject(body)
            if (json.optBoolean("ok", false)) {
                val meta = json.getJSONObject("meta").let {
                    ApiMeta(it.optBoolean("cached"), it.optString("timestamp"), it.optString("version"))
                }
                ApiResult.Success(json.getJSONObject("data"), meta)
            } else {
                val err = json.getJSONObject("error")
                ApiResult.Error(err.optString("code", "INTERNAL_ERROR"), err.optString("message"), err.optInt("status", 500))
            }
        } catch (e: IOException) { ApiResult.NetworkError(e) }
    }

    suspend fun ping() = get("/api/v1/ping")
    suspend fun me() = get("/api/v1/me")
    suspend fun getBibles() = get("/api/v1/bibles")
    suspend fun getBible(bibleId: Int) = get("/api/v1/bibles/$bibleId")
    suspend fun getBooks(bibleId: Int) = get("/api/v1/bibles/$bibleId/books")
    suspend fun getBook(bibleId: Int, bookUsfm: String) = get("/api/v1/bibles/$bibleId/books/$bookUsfm")
    suspend fun getChapters(bibleId: Int, bookUsfm: String) = get("/api/v1/bibles/$bibleId/books/$bookUsfm/chapters")
    suspend fun getChapter(bibleId: Int, chapterUsfm: String) = get("/api/v1/bibles/$bibleId/chapters/$chapterUsfm")
    suspend fun getVerseList(bibleId: Int, chapterUsfm: String) = get("/api/v1/bibles/$bibleId/chapters/$chapterUsfm/verses")
    suspend fun getVerse(bibleId: Int, verseUsfm: String) = get("/api/v1/bibles/$bibleId/verses/$verseUsfm")
    suspend fun getPassage(bibleId: Int, reference: String) = get("/api/v1/bibles/$bibleId/passages/$reference")
    suspend fun getBibleIndex(bibleId: Int) = get("/api/v1/bibles/$bibleId/index")
    suspend fun getLanguages() = get("/api/v1/languages")
    suspend fun getLanguage(langId: Int) = get("/api/v1/languages/$langId")
    suspend fun getVotdToday() = get("/api/v1/votd/today")
    suspend fun getVotdYear(year: Int) = get("/api/v1/votd/$year")
    suspend fun getVotdDay(year: Int, day: Int) = get("/api/v1/votd/$year/$day")
}

// Contoh penggunaan di ViewModel:
//
// class HomeViewModel : ViewModel() {
//     private val api = BibleApiClient()
//     fun loadVotd() = viewModelScope.launch {
//         when (val r = api.getVotdToday()) {
//             is ApiResult.Success -> { /* r.data.getString("text") */ }
//             is ApiResult.Error   -> { if (r.code == "UNAUTHORIZED") { /* redirect login */ } }
//             is ApiResult.NetworkError -> { /* tampilkan pesan offline */ }
//         }
//     }
// }

File: examples/ios/BibleApiClient.swift
// ============================================================
// Bible Mirror API — iOS Swift Client
// ============================================================
// SPM: firebase-ios-sdk → FirebaseAuth
// ============================================================

import Foundation
import FirebaseAuth

enum BibleApiConfig {
    // Ganti dengan URL Worker Anda setelah deploy
    static let baseURL = "https://bible-mirror-api.<your-subdomain>.workers.dev"
}

struct ApiMeta: Decodable { let cached: Bool; let timestamp: String; let version: String }
struct ApiSuccess<T: Decodable>: Decodable { let ok: Bool; let data: T; let meta: ApiMeta }
struct ApiErrorDetail: Decodable { let code: String; let message: String; let status: Int }
struct ApiErrorResponse: Decodable { let ok: Bool; let error: ApiErrorDetail }

enum ApiResult<T> {
    case success(data: T, meta: ApiMeta)
    case failure(code: String, message: String, status: Int)
    case networkError(Error)
}

actor BibleApiClient {
    private let session: URLSession
    private let decoder: JSONDecoder

    init(session: URLSession = .shared) {
        self.session = session
        self.decoder = JSONDecoder()
    }

    private func getIdToken() async throws -> String {
        guard let user = Auth.auth().currentUser else { throw URLError(.userAuthenticationRequired) }
        return try await user.getIDToken()
    }

    private func get<T: Decodable>(path: String, as type: T.Type) async -> ApiResult<T> {
        do {
            let token = try await getIdToken()
            guard let url = URL(string: "\(BibleApiConfig.baseURL)\(path)") else {
                return .failure(code: "BAD_REQUEST", message: "Invalid URL", status: 400)
            }
            var req = URLRequest(url: url)
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            req.setValue("application/json", forHTTPHeaderField: "Accept")
            req.timeoutInterval = 15
            let (data, _) = try await session.data(for: req)
            if let success = try? decoder.decode(ApiSuccess<T>.self, from: data) {
                return .success(data: success.data, meta: success.meta)
            }
            if let err = try? decoder.decode(ApiErrorResponse.self, from: data) {
                return .failure(code: err.error.code, message: err.error.message, status: err.error.status)
            }
            return .failure(code: "INTERNAL_ERROR", message: "Unexpected response", status: 500)
        } catch { return .networkError(error) }
    }

    func ping() async -> ApiResult<[String: String]> { await get(path: "/api/v1/ping", as: [String: String].self) }
    func me() async -> ApiResult<[String: String]> { await get(path: "/api/v1/me", as: [String: String].self) }
    func getBibles() async -> ApiResult<[String: AnyDecodable]> { await get(path: "/api/v1/bibles", as: [String: AnyDecodable].self) }
    func getBible(bibleId: Int) async -> ApiResult<[String: AnyDecodable]> { await get(path: "/api/v1/bibles/\(bibleId)", as: [String: AnyDecodable].self) }
    func getBooks(bibleId: Int) async -> ApiResult<[String: AnyDecodable]> { await get(path: "/api/v1/bibles/\(bibleId)/books", as: [String: AnyDecodable].self) }
    func getChapters(bibleId: Int, bookUsfm: String) async -> ApiResult<[String: AnyDecodable]> { await get(path: "/api/v1/bibles/\(bibleId)/books/\(bookUsfm)/chapters", as: [String: AnyDecodable].self) }
    func getChapter(bibleId: Int, chapterUsfm: String) async -> ApiResult<[String: AnyDecodable]> { await get(path: "/api/v1/bibles/\(bibleId)/chapters/\(chapterUsfm)", as: [String: AnyDecodable].self) }
    func getVerseList(bibleId: Int, chapterUsfm: String) async -> ApiResult<[String: AnyDecodable]> { await get(path: "/api/v1/bibles/\(bibleId)/chapters/\(chapterUsfm)/verses", as: [String: AnyDecodable].self) }
    func getVerse(bibleId: Int, verseUsfm: String) async -> ApiResult<[String: AnyDecodable]> { await get(path: "/api/v1/bibles/\(bibleId)/verses/\(verseUsfm)", as: [String: AnyDecodable].self) }
    func getPassage(bibleId: Int, reference: String) async -> ApiResult<[String: AnyDecodable]> { await get(path: "/api/v1/bibles/\(bibleId)/passages/\(reference)", as: [String: AnyDecodable].self) }
    func getLanguages() async -> ApiResult<[String: AnyDecodable]> { await get(path: "/api/v1/languages", as: [String: AnyDecodable].self) }
    func getVotdToday() async -> ApiResult<[String: AnyDecodable]> { await get(path: "/api/v1/votd/today", as: [String: AnyDecodable].self) }
    func getVotdDay(year: Int, day: Int) async -> ApiResult<[String: AnyDecodable]> { await get(path: "/api/v1/votd/\(year)/\(day)", as: [String: AnyDecodable].self) }
}

// Helper untuk decode JSON dinamis
struct AnyDecodable: Decodable {
    let value: Any
    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if let v = try? c.decode(String.self) { value = v }
        else if let v = try? c.decode(Int.self) { value = v }
        else if let v = try? c.decode(Double.self) { value = v }
        else if let v = try? c.decode(Bool.self) { value = v }
        else if let v = try? c.decode([String: AnyDecodable].self) { value = v }
        else if let v = try? c.decode([AnyDecodable].self) { value = v }
        else { value = NSNull() }
    }
}
